src/main/java/org/dependencytrack/tasks/OSVDownloadTask.java

```diff
@@ -0,0 +1,277 @@
+    // calculate severity of vulnerability on priority-basis (database, ecosystem)
+    public Severity calculateOSVSeverity(OSVAdvisory advisory) {
+
+        // derive from database_specific cvss v3 vector if available
+        if(advisory.getCvssV3Vector() != null) {
+            Cvss cvss = Cvss.fromVector(advisory.getCvssV3Vector());
+            Score score = cvss.calculateScore();
+            return normalizedCvssV3Score(score.getBaseScore());
+        }
+        // derive from database_specific cvss v2 vector if available
+        if (advisory.getCvssV2Vector() != null) {
+            Cvss cvss = Cvss.fromVector(advisory.getCvssV2Vector());
+            Score score = cvss.calculateScore();
+            return normalizedCvssV2Score(score.getBaseScore());
+        }
+        // get database_specific severity string if available
+        if (advisory.getSeverity() != null) {
+            if (advisory.getSeverity().equalsIgnoreCase("CRITICAL")) {
+                return Severity.CRITICAL;
+            } else if (advisory.getSeverity().equalsIgnoreCase("HIGH")) {
+                return Severity.HIGH;
+            } else if (advisory.getSeverity().equalsIgnoreCase("MODERATE")) {
+                return Severity.MEDIUM;
+            } else if (advisory.getSeverity().equalsIgnoreCase("LOW")) {
+                return Severity.LOW;
+            }
+        }
+        // get largest ecosystem_specific severity from its affected packages
+        if (advisory.getVulnerabilities() != null) {
+            List<Integer> severityLevels = new ArrayList<>();
+            for (OSVVulnerability vuln : advisory.getVulnerabilities()) {
+                severityLevels.add(vuln.getSeverity().getLevel());
+            }
+            Collections.sort(severityLevels);
+            Collections.reverse(severityLevels);
+            return getSeverityByLevel(severityLevels.get(0));
+        }
+        return Severity.UNASSIGNED;
+    }
+
+    public Vulnerability.Source extractSource(String vulnId) {
+
+        final String sourceId = vulnId.split("-")[0];
+        switch (sourceId) {
+            case "GHSA": return Vulnerability.Source.GITHUB;
+            case "CVE": return Vulnerability.Source.NVD;
+            default: return Vulnerability.Source.GOOGLE;
+        }
+    }
+
+    private VulnerableSoftware mapVulnerabilityToVulnerableSoftware(final QueryManager qm, final OSVVulnerability vuln) {
+
+        String versionStartIncluding = vuln.getLowerVersionRange();
+        String versionEndExcluding = vuln.getUpperVersionRange();
+
+        final String purl = vuln.getPurl();
+        if (purl == null) return null;
+
+        VulnerableSoftware vs = qm.getVulnerableSoftwareByPurl(vuln.getPurl(), versionEndExcluding, versionStartIncluding);
+        if (vs != null) {
+            return vs;
+        }
+        vs = new VulnerableSoftware();
+        vs.setVulnerable(true);
+        vs.setPurlType(vuln.getPackageEcosystem());
+        vs.setPurl(vuln.getPurl());
```

Comment:
Sorry I didn't spot this earlier, only noticed it while testing right now: Instead of setting the entire PURL, you need to set the PURL components separately, e.g.

```java
vs.setPurlType(vuln.getPackageEcosystem());
vs.setPurlNamespace(...);
vs.setPurlName(...);
```

Otherwise the `InternalAnalysisTask` won't pick them up as of now. Here's the code used by `InternalAnalysisTask` to query for applicable `VulnerableSoftware`s:

https://github.com/DependencyTrack/dependency-track/blob/a3b42a6fe4d19fc91122887dba0841f26fa7c7c4/src/main/java/org/dependencytrack/persistence/VulnerableSoftwareQueryManager.java#L197-L220

Here's how it's done for GHSA:

https://github.com/DependencyTrack/dependency-track/blob/a3b42a6fe4d19fc91122887dba0841f26fa7c7c4/src/main/java/org/dependencytrack/tasks/GitHubAdvisoryMirrorTask.java#L283-L285